### PR TITLE
 display progress and allow cancelling when downloading server

### LINKED ISSUE #263: should show progress when downloading a runtime
steps:
1. EXEC: pick "Download Runtime" in the context menu of the "Servers" view
2. EXEC: pick a runtime (ex. wildfly-11) & provide answers to all questions (ex. accept the license etc.)
3. ASSERT: you get an info dialog that the runtime is being downloaded
image
4. ASSERT: at some point the info dialog disappears

Result:
From then on you have no idea when the download will be finished, if it's downloading successfully, what's then going to happen etc.

Expected result:
We should provide some progress, ex. use the progress bar on top of the explorer to show that the download really happens, that it's progressing etc.

### PR DESCRIPTION
**Pull request by adietish**
fixes #263, #266

### CODE REVIEW

File: src/jobprogress.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  private isJob(job: Protocol.JobHandle): boolean {
+    return this.job.id === job.id;
+  }
+
+  private restartTimeout() {
+    this.clearTimeout();
+    this.setTimeout(this.reject);
+  }
+
+  private setTimeout(reject: (reason?: any) => void) {
```
Comment (jrichter1) [inline src/jobprogress.ts]: the argument is not being used

File: src/jobprogress.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+  private onJobRemoved(jobRemoved: Protocol.JobRemoved) {
+      if (!this.isJob(jobRemoved.handle)) {
+        return;
+      }
+      this.clearTimeout();
+      this.resolve();
+  }
+
+  private onCancel() {
```
Comment (jrichter1) [inline src/jobprogress.ts]: this might look better as an async method - not crucial but nicer :)

**Comment by adietish** [comment]
@jrichter1 addressed your complaints, added tests. Please re-review. Thx

File: test/jobprogress.test.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    test('onJobAdded notification should open window with progress', () => {
+      // given
+      JobProgress.create(stubs.client);
+
+      // when
+      callOnJobAddedListenerWith(job, stubs.incoming.onJobAdded);
+
+      // then
+      expect(stubs.incoming.onJobAdded).calledOnce;
```
Comment (jrichter1) [inline test/jobprogress.test.ts]: you can extend the client call expectations to also verify the parameters
```calledOnceWith(job)```

**Comment by jrichter1** [comment]
Also please check the formatting - the indentation is all over the place. Lets just settle on 4 spaces, as the rest of the files seem to be using that.

File: test/jobprogress.test.ts (hunk 3 above)
Comment (adietish) [reply test/jobprogress.test.ts]: @jrichter1 you're right in general, but here onJobAdded is called with the listener to get informed of a new job. The assertion should make sure that jobProgress registered its (internal) listener, not a specific one.

Comment (jrichter1) [reply test/jobprogress.test.ts]: I admit, this line is about the worst example I could have picked, but I hope you get the idea :)

File: src/serverExplorer.ts [hunk 4]
```diff
@@ -161,7 +161,7 @@ export class ServersViewTreeDataProvider implements TreeDataProvider< Protocol.S
                 trace: ``,
                 plugin: ``,
                 ok: false,
-                message: `Canceled by user`
+                message: `Cannot add server, user didnt select a server runtime folder.`
```
[the remark this answers is not in the document]
Comment (adietish) [reply src/serverExplorer.ts]: @dgolovin this code is outdated, once this PR is rebase there won't be this bit. It was removed bcs we shouldn't error if a user cancels an operation ;)

**Pull request merged**